gateway/ln-gateway/src/lightning/lnd.rs

```diff
@@ -847,6 +831,93 @@ impl ILnRpcClient for GatewayLndClient {
     }
 }
 
+#[derive(Debug)]
+struct PrivatePaymentResult {
+    preimage: Option<Vec<u8>>,
+    last_failed_reason: Option<String>,
+}
+
+async fn attempt_private_payments(
+    routes: QueryRoutesResponse,
+    invoice: PrunedInvoice,
+    client: &mut LndClient,
+) -> Result<PrivatePaymentResult, LightningRpcError> {
+    let total_amt_msat: i64 =
+        invoice
+            .amount
+            .msats
+            .try_into()
+            .map_err(|error| LightningRpcError::FailedPayment {
+                failure_reason: format!("amount exceeds valid LND amount ranges {error:?}"),
+            })?;
+
+    let mut preimage: Option<Vec<u8>> = None;
+    let mut last_failed_reason: Option<String> = None;
+
+    for mut route in routes.routes {
+        // skip any empty routes
+        if route.hops.is_empty() {
+            continue;
+        }
+
+        // add mpp record to route
+        let last = route.hops.last_mut().expect("must have at least one hop");
+
+        last.mpp_record = Some(MppRecord {
+            payment_addr: invoice.payment_secret.to_vec(),
+            total_amt_msat,
+        });
+
+        trace!("Attempting to send payment to route {route:?}");
+
+        debug!("LND payment request sent for invoice {invoice:?}, waiting for payment status...");
+
+        let htlc_attempt = client
+            .router()
+            .send_to_route_v2(SendToRouteRequest {
+                payment_hash: invoice.payment_hash.to_byte_array().to_vec(),
+                route: Some(route),
+                skip_temp_err: false,
+            })
+            .await
+            .map_err(|status| {
+                info!("LND payment request failed for invoice {invoice:?} with {status:?}");
+                LightningRpcError::FailedPayment {
+                    failure_reason: format!("Failed to make outgoing payment {status:?}"),
+                }
+            });
+
+        match htlc_attempt.map(Response::into_inner) {
+            Ok(payment) => match payment.status() {
+                HtlcStatus::Failed => {
+                    info!("LND payment failed for invoice {invoice:?} with {payment:?}");
+                    if let Some(failure) = payment.failure {
+                        last_failed_reason = Some(failure.code().as_str_name().to_string());
+                    }
+                }
+                HtlcStatus::InFlight => {
+                    debug!("LND payment is inflight");
+                    continue;
```

Comment:
Is each `route` in this for loop a separate complete route to the destination? I'm trying to understand why we want to continue here when a payment is inflight. This will cause `send_to_route_v2` RPC to be called again with a different route right?

Reply:
ah yeah I thought I was processing a stream, need to modify here